src/python/pants/backend/k8s/goals/deploy.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2024 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+
+from pants.backend.docker.goals.package_image import DockerPackageFieldSet
+from pants.backend.k8s.k8s_subsystem import K8sSubsystem
+from pants.backend.k8s.kubectl_subsystem import Kubectl
+from pants.backend.k8s.target_types import (
+    K8sBundleContextField,
+    K8sBundleDependenciesField,
+    K8sBundleSourcesField,
+    K8sSourceField,
+)
+from pants.core.goals.deploy import DeployFieldSet, DeployProcess
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.engine.addresses import UnparsedAddressInputs
+from pants.engine.env_vars import EnvironmentVars, EnvironmentVarsRequest
+from pants.engine.fs import MergeDigests, Snapshot
+from pants.engine.internals.native_engine import Digest
+from pants.engine.platform import Platform
+from pants.engine.process import InteractiveProcess, Process, ProcessCacheScope
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    DependenciesRequest,
+    HydratedSources,
+    HydrateSourcesRequest,
+    SourcesField,
+    Targets,
+)
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DeployK8sBundleFieldSet(DeployFieldSet):
+    required_fields = (
+        K8sBundleSourcesField,
+        K8sBundleContextField,
+        K8sBundleDependenciesField,
+    )
+    sources: K8sBundleSourcesField
+    context: K8sBundleContextField
+    dependencies: K8sBundleDependenciesField
+
+
+@dataclass(frozen=True)
+class KubectlApply:
+    paths: tuple[str, ...]
+    input_digest: Digest
+    platform: Platform
+    env: FrozenDict[str, str] | None = None
+    context: str | None = None
+
+
+@rule(desc="Run k8s deploy process", level=LogLevel.DEBUG)
+async def run_k8s_deploy(
+    field_set: DeployK8sBundleFieldSet,
+    kubectl: Kubectl,
+    k8s_subsystem: K8sSubsystem,
+    platform: Platform,
+) -> DeployProcess:
+    context = field_set.context.value
+    if context is None:
```

Comment:
I think I've missed something in the logic here. Why do we force the `field_set.context` to have a value even when `kubectl.pass_context` is `False`?

Reply:
The idea is that setting `context` on a target is required. This is because we had recurring issues because somebody forgot to set the correct context on the target and it got deployed to a different context by accident. However, we've got CI agents running in the cluster itself, so they can only deploy k8s objects to the specific cluster specified with `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT`, that's why we need to disable `--context` argument in `pants.ci.toml`.

I understand this is kinda custom setup, so people might need some other behavior. However, in general, I think context requirement is a good thing that can prevent misdeployments. I can probably move this context validation to a small linter and check `context` field there, wdyt?

Reply:
It makes sense to have a failsafe to prevent deploying things in the wrong clusters. I was thinking of how the Helm backend does it. It has an optional `HelmDeploymentNamespaceField` and no field for context.

I think we can just add a line in the docs explaining the intended workflow. Something like
> To prevent accidentally deploying kubernetes manifests to the wrong cluster, the `context` field is required on `k8s_bundle`s for deployment.
> For deploying the same `k8s_bundle` to multiple contexts, consider using `parametrize` like `k8s_bundle(context=parametrize("stg", "prd"))`
> For CI agents which will only have access to a single context, set the `kubectl.pass_context` to `false` to have them use their default context.

Reply:
> I was thinking of how the Helm backend does it.

We have the exact same problem in helm backend. I had to vendor it and keep maintaining the patch that adds the context requirement

Reply:
I have 8 kubernetes clusters I manage. I have some simple scripts to run helm and deploy kubernetes resources. Eventually I would love to switch to pants. One of the key features in my scripts is a context assertion to make sure I'm in the correct context because deploying to the wrong cluster would be very bad. Namespaces are also important, but not as important as the context.

Having this context protection feature in pants looks fantastic.